### ISSUE TITLE
manual addition of wildcard (*) URLs for username-only detection in site preferences broken on firefox

### ISSUE THREAD

**Issue by brad2014** [keepassxreboot/keepassxc-browser#1612]
## Expected Behavior

The login page for https://qbo.intuit.com redirects to a random subdomain host on each visit, for example:  https://c19.qbo.intuit.com/qbo19/login.  That login page only has a username field, with the password field on the next page.

On Brave (Browser: Chrome/Chromium 100.0.4896.127), in KeepassXC plugin's Site Preferences, I can manually add the URL `https://*.qbo.intuit.com/*` and enable username-only detection for the entire subdomain.  This allows me to use KeePassXC-Browser autofill no matter which subdomain host processes my login.

On Firefox, it should do the same.

## Current Behavior

Unfortunately, on Firefox (unlike Chrome), I get a validation error when I click "+ Add" that prevents me from adding the wildcard URL:

<img width="509" alt="Site Preferences Error" src="https://user-images.githubusercontent.com/6502462/164117839-4940f6ae-0228-4e47-be19-ff8bc9555f00.png">
## Possible Solution

Unknown.

## Steps to Reproduce (for bugs)

1. Visit https://qbo.intuit.com, and get redirected to the login page.
2. Try to autofill the username into the username field.
3. Try to manually add the site preference to enable username-only detection for`https://*.qbo.intuit.com/*` so it works regardless of which host one gets redirected to.
4. Notice error message on pressing "+ Add" button.

## Debug info

KeePassXC - 2.7.1
KeePassXC-Browser - 1.7.12
Operating system: MacIntel
Browser: Mozilla Firefox 99.0

**Comment by droidmonkey** [comment]
Remove the /*

**Comment by varjolintu** [comment]
The field's `ValidationState` differs:
Edge:
<img width="189" alt="edge" src="https://user-images.githubusercontent.com/24570482/164147006-215fe4a2-36d9-4d99-9037-b7bb5c0af749.png">

Firefox:
<img width="153" alt="firefox" src="https://user-images.githubusercontent.com/24570482/164147023-ee03378f-b522-4903-8530-d4de42e309a0.png">

`typeMismatch` means Firefox doesn't accept that URL as a valid one, but Chromium-based browsers do. We could probably check the validity against `patternMismatch` but the field would still show an error in real-time.

**Comment by a2kolbasov** [comment]
> `typeMismatch` means Firefox doesn't accept that URL as a valid one, but Chromium-based browsers do. We could probably check the validity against `patternMismatch` but the field would still show an error in real-time.

Could you please explain, if there are any reasons why using `<input type="text" pattern=...` might not be advisable? In this case, browsers will not check the input through an internal pattern.

**Comment by varjolintu** [comment]
> > `typeMismatch` means Firefox doesn't accept that URL as a valid one, but Chromium-based browsers do. We could probably check the validity against `patternMismatch` but the field would still show an error in real-time.
> 
> Could you please explain, if there are any reasons why using `<input type="text" pattern=...` might not be advisable? In this case, browsers will not check the input through an internal pattern.

That's one possible solution I have been thinking of.

**Comment by rnc** [comment]
Doesn't this mean wildcards on Firefox in site preferences for disabling features *or* enabling features (like username only detection) are broken i.e. this issue's title is too narrow and the scope is far wider? I wanted to add an exclusion for all my work company domains (i.e. *.mycompany.com and disable all features) but I can't do that on firefox.

**Comment by a2kolbasov** [comment]
> Doesn't this mean wildcards on Firefox in site preferences for disabling features _or_ enabling features (like username only detection) are broken i.e. this issue's title is too narrow and the scope is far wider? I wanted to add an exclusion for all my work company domains (i.e. *.mycompany.com and disable all features) but I can't do that on firefox.

No. In the site preferences open the DevTools (<kbd>F12</kbd>) and execute
```javascript
document.getElementById("manualUrl").type = "text";
```
This will disable the browser's validator, and you’ll be able to enter a URL with wildcards. The problem is only in the text field.
